dimod/generators/mimo.py

```diff
@@ -0,0 +1,1085 @@
+
+        modulation: Modulation. Supported values are 'BPSK', 'QPSK', '16QAM',
+            '64QAM', and '256QAM'.
+
+        channel_power: Channel power. By default, proportional to the number
+            of transmitters.
+
+        random_state: Seed for a random state or a random state.
+
+    Returns:
+        Four-tuple of received signals (``y``), transmitted symbols (``v``),
+        channel noise, and random_state, where ``y`` is a column vector of length
+        equal to the rows of ``F``.
+    """
+
+    num_receivers = F.shape[0]
+    num_transmitters = F.shape[1]
+
+    random_state = _make_random_state(random_state)
+
+    bits_per_transmitter, amps, constellation_mean_power = _constellation_properties(modulation)
+
+    if transmitted_symbols is not None:
+        if modulation == 'BPSK' and np.iscomplex(transmitted_symbols).any():
+            raise ValueError(f"BPSK transmitted signals must be real")
+        if modulation != 'BPSK' and np.isreal(transmitted_symbols).any():
+            raise ValueError(f"Quadrature transmitted signals must be complex")
+    else:
+        if type(random_state) is not np.random.mtrand.RandomState:
+            random_state = np.random.RandomState(random_state)
+
+        quadrature = False if modulation == 'BPSK' else True
+        transmitted_symbols = _create_transmitted_symbols(
+                num_transmitters, amps=amps, quadrature=quadrature, random_state=random_state)
+
+    if SNRb <= 0:
+       raise ValueError(f"signal-to-noise ratio must be positive. SNRb={SNRb}")
+
+    if SNRb == float('Inf'):
+       y = np.matmul(F, transmitted_symbols)
+    elif channel_noise is not None:
+        y = channel_noise + np.matmul(F, transmitted_symbols)
+    else:
+        # Energy_per_bit:
+        if channel_power == None:
+            #Assume proportional to num_transmitters; i.e., every channel 
+            # component is RMSE 1 and 1 bit
+            channel_power = num_transmitters
+
+        #Eb is the same for QPSK and BPSK
+        # Eb/N0 = SNRb (N0 = 2 sigma^2, the one-sided PSD ~ kB T at antenna)
+        # SNRb and Eb, together imply N0
+        Eb = channel_power * constellation_mean_power / bits_per_transmitter 
+        N0 = Eb / SNRb
+        # Noise is complex by definition, hence 1/2 power in real & complex parts
+        sigma = np.sqrt(N0 / 2) 
+
+        # Channel noise of covariance N0*I_{NR}. Noise is complex by definition, 
+        # although for real channel and symbols we need only worry about 
+        # real part:
+        channel_noise = sigma*(random_state.normal(0, 1, size=(num_receivers, 1)) \
+            + 1j*random_state.normal(0, 1, size=(num_receivers, 1)))
+        if modulation == 'BPSK' and np.isreal(F).all():
+            channel_noise = channel_noise.real
+
+        y = channel_noise + np.matmul(F, transmitted_symbols)
+
+    return y, transmitted_symbols, channel_noise, random_state
+
+def spin_encoded_mimo(modulation: str, 
```

Comment:
As this method returns a BQM that has a `.vartype` attribute, IMO we should just call this `encoded_mimo()`.

Reply:
I renamed to `comp` and `mimo`, which required renaming the module, which is more in keeping with other generator names (encoding as a BQM is implied).

Reply:
No strong preference.
The natural statement of the mimo problem is a quadratic function on constellation states. This is how the problem would be presented to the hybrid solver (for example). In light of Alex's observation I'd suggest bqm_encoded_mimo, but if thats not standard I dont mind. Note that there are multiple ways to encode (gray coding etc), but the simplest form is used here.

Reply:
Closed as part of [this discussion](https://github.com/dwavesystems/dimod/pull/1357#discussion_r1310526351)